### ISSUE TITLE
suggestion: a decoder/assertion library

### ISSUE THREAD

**Issue by thefliik** [denoland/deno_std#491]
Something I've found to be a recurring need: taking an `unknown` input, asserting that it has a specific shape, and either returning a response that is correctly typed in typescript or returning an error.

You could call this a decoder library or an assertion library, but would adding a package to deno_std that deals with this be appropriate?

As an example, recently I started using a small but helpful library for this purpose [`ts.data.json`](https://github.com/joanllenas/ts.data.json) (I'm not affiliated with the project). Its main export is a `Decoder` class.

From the source:

```ts
const string: Decoder<string> = new Decoder<string>((json: any) => {
    if (typeof json === 'string') {
      return ok<string>(json);
    } else {
      return err<string>($JsonDecoderErrors.primitiveError(json, 'string'));
    }
  })
```

This decoder can be used to both assert that an `unknown` is a `string`, and properly type the result. **Importantly**, it is also composable with other decoders (example below).

I'm not wedded to the `ts.data.json` api (though I like it), but, in general, would something along these lines be appropriate to add to deno_std? This seems like something that many projects could benefit from.

A more complex example:

```ts
interface IUser {
  firstname: string;
  lastname: string;
};

const userDecoder = JsonDecoder.object<IUser>(
  {
    firstname: JsonDecoder.string,
    lastname: JsonDecoder.string
  },
  'User'
);

const unknownJSON = {
  firstname: 1,
  lastname: 'Smith'
};

// This would actually throw a type error because the type of `unknownJSON`
// is known to violate the type that `userDecoder` expects, but lets assume that
// `unknownJSON` is really unknown
const result = userDecoder.decode(unknownJSON)

if (result instanceof Err) {
  throw new Error(
    result.err // error message
  );
}

// properly typed as IUser
const user = result.value
```

**Comment by thefliik** [comment]
Beyond simply providing a unified API for (what I think is) a common programming task, a decoder library is generally going to be related to an application's security, so I think it could benefit from being maintained in `deno_std`.

**Comment by thefliik** [comment]
#484 could be related (in that a generalized decoder API would be able to handle the assertions described in that issue).

The main advantage of providing a decoder API, instead of using individual `assert()` functions, would be composability.

**Comment by j-f1** [comment]
There’s [an implementation of this](https://github.com/nvie/decoders) in Node.js land, but unfortunately it’s Flow-native rather than TS-native.

**Comment by thefliik** [comment]
@j-f1 that's a great example. My hope in opening this issue was to see if a pull request would be accepted, as well as start a discussion of what an API might look like.

`ts.data.json` returns either an `Ok<T>` (on success) or `Err<T>` (on failure) object from a Decoder. It looks like the `decoders` library you've linked to throws an error on failure and simply returns the value on success.

At first I was a little annoyed by the `Ok` and `Err` values being returned from a decoder, but on reflection I thought that throwing an error would probably be heavy handed for a decoder error given that a decoder error is not *exceptional* but rather one of two expected outcomes. Throwing an error would also prevent the possibility of attempting to decode other values of, say, an array decoder and possibly returning an array of errors (note: I'm not sure if that functionality would be desired).

Do you have an opinion here? Do others?

Seperately, what decoders should be included in this hypothetical package? For the most part, I like the defaults included in the `decoders` library and would add, at minimum, an `allOf` decoder and a `lazy` decoder (for recursive data structures).

## Proposed decoders:

Decoders should be capable of decoding a promise which returns a value or, in the case of higher order decoders, a promise which returns a decoder.

### Primitives

1. **number**
     - Returns a decoder capable of decoding finite (!) numbers (integer or float values). This means that values like NaN, or positive and negative Infinity are not considered valid numbers.
2. **integer**
     - Like number, but only decodes values that are whole numbers.
3. **string**
     - Returns a decoder capable of decoding string values.
4. **regex**
     - Returns a decoder capable of decoding string values that match the given regular expression.
5. **boolean**
     - Returns a decoder capable of decoding boolean values.
6. **null**
     - Returns a decoder capable of decoding the constant value `null` (same as `constant(null)`).
7. **undefined**
     - Returns a decoder capable of decoding the constant value `undefined` (same as `constant(undefined)`).
8. **constant**
     - Returns a decoder capable of decoding just the given constant value.
9. **hardcoded** (don't love the name. Could be called `exactly` as well, which is more intuitive?)
     - Returns a decoder that will always return the provided value without looking at the input.
10. **never**
     - Returns a decoder that will always fail with the given error messages, no matter what the input. 
11. **any**
     - Returns a decoder that will always succeed, passing through any input value (could also be called `unknown` but I think any is more intuitive).
     - Should this be typed as `any` or `unknown`?
12. **instanceof**
    - Returns a decoder capable of decoding values that are instances of the given class.

### Objects

1. **array**
   - If provided a decoder for type T, returns a decoder capable of decoding an array of T's
2. **tuple**
   - Returns a decoder capable of decoding a 2-tuple of (T1, T2)'s, provided that you already have a decoder for T1 and T2. A tuple is like an Array, but the number of items in the array is fixed (two) and their types don't have to be homogeneous. This decoder should have overloads accepting tuples of increasing size up to some to-be-determined amount.
3. **object**
   - Returns a decoder capable of decoding objects of a given shape, provided that you already have decoders for all values in the mapping.
4. **exactObject**
   - Like object, but will fail if there are superfluous keys in the input data.
5. **dictionary**
   - If provided a decoder for type T, returns a decoder capable of decoding every value of an object no matter the key.

### Compositional

1. **optional**
     - If provided a decoder for type T, returns a decoder capable of decoding either a value of type T, or undefined (shorthand for `oneOf(aDecoder, undefinedDecoder)`).
2. **nullable**
     - If provided a decoder for type T, returns a decoder capable of decoding either a value of type T, or null (shorthand for `oneOf(aDecoder, nullDecoder)`).
3. **maybe**
   - If provided a decoder for type T, returns a decoder capable of decoding either a value of type T, or null, or undefined (shorthand for `oneOf(aDecoder, nullDecoder, undefinedDecoder)`).
4. **oneOf**
   - Accepts a spread of decoders and succeeds if the provided value can be successfully decoded by any of the provided decoders.
5. **allOf**
   - Accepts a spread of decoders and succeeds if the provided value can be successfully decoded by all of the provided decoders, in order. The output of one decoder is passed as input to the next decoder. It returns Ok with the last successful decoded value or Err if any decoder fails.
6. **lazy**
   - Decoder for recursive data structures. Accepts a function which returns a decoder.

### Misc other decoders which *could* be included

This list could go on for a while and I'm unsure what would be appropriate for a standard library. I'm tempted to say simply cut the list short and not include any `misc` decoders (let other libraries fill in the gap), but all of the following seem like broadly useful inclusions....

1. email
2. date
3. time
4. datetime ([ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) ?)
5. url
6. length (?)
7. greaterThan (?)
8. lessThan (?)
9. uuid (?) <-- perhaps this should be included in the `deno_std` uuid library instead.

**Comment by j-f1** [comment]
IMO `allOf` should be called something like `chain` to clarify that it passes the result of one decoder to the next one.
How about `value` for the `hardcoded` decoder?
For `tuple`, here’s a type definition that will work for an arbitrarily-large number of types:

```ts
declare function tuple<Tuple extends any[]>(
  ...args: { [I in keyof Tuple]: Decoder<Tuple[I]> }
): Decoder<Tuple>
```

**Comment by thefliik** [comment]
@j-f1 

> IMO `allOf` should be called something like `chain` to clarify that it passes the result of one decoder to the next one.

That's fair. I chose the name as the counterpart to `oneOf`, but I think your suggestion is better. I've updated my previous suggestion to `chain`.

> How about `value` for the `hardcoded` decoder?

Initially I liked the suggestion, but on second thought I think it's too similar to the `constant` decoder (having both `value` and `constant` seems confusing)? Really, the two decoders are very similar. What about `insertConstant` instead?
